### PR TITLE
WIP: Login method api

### PR DESCRIPTION
**Pull request by jess-sheneberger**
## Related issue

There is no related issue.

## Proposed changes

In my app I want to allow users to register via one of several methods: two OIDC providers or a password.  When they return after sign-up, I want to provide a login experience that doesn't require them to remember how they registered. I want them to enter their username or email address and for the UI to start the correct OIDC provider flow or prompt for their password using the same method they registered with.

To make this possible, I've put together this login method API. It looks up the given email address and returns either the ID of the OIDC provider that should be used, the "password" method.  My app UI is able to interpret the result and either launch the correct OIDC flow or prompt for a password.

I also have a requirement that users who login via password need to have a username rather just than an email address.  Login via email address is fine for users who login via OIDC.

I understand this is a potential account enumeration attack vector, so maybe it should be turned on with a configuration option and left off for most users.

I'm open to different/better approaches to accomplish this. I'm marking it WIP because I implemented what seemed to me to be the most straightforward way, but maybe there is a different approach that would work better.

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security. vulnerability, I
      confirm that I got green light (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [ ] I have added tests that prove my fix is effective or that my feature
      works.
- [ ] I have added or changed [the documentation](docs/docs).

### CODE REVIEW

**Comment by aeneasr** [comment]
If we put this into the admin endpoint I think it would be mergeable. It could also allow UIs to have a UI indicator if a username is already taken!

**Comment by jess-sheneberger** [comment]
Sure, I can do that.

Right now in my configuration the admin endpoint isn't available to the client, so we'd need some way to expose this one admin API method to the client, right?

**Comment by aeneasr** [comment]
Yes, you would usually do that in your routing / reverse proxy set up - exposing this specific endpoint to the public internet

**Comment by aeneasr** [comment]
So should we keep this PR open or is it something that you were able to work around / do not intend on continue working on?

**Comment by jess-sheneberger** [comment]
Let's keep it open, I'm still working on it.  I've got it moved over to the admin endpoint but I'm wrestling with my reverse proxy ATM so I can't test it yet.

**Comment by jess-sheneberger** [comment]
OK, it's on the admin endpoint now. What else should I do before this can be merged?

File: identity/handler.go
Comment (aeneasr) [inline identity/handler.go:75]: If I understand correctly you want to check if a user is signed up using the provided credentials, correct? In that case it would probably make sense to have a `POST` request which accepts a payload like

```
{
  "identifier": "...",
  "method": "password" # or "oidc"
}
```

and returns a `true` or `false` if it was found or not found? I think this would make the logic below much easier!

If the `method` is omitted - which means we want to discover the method - we could still implement the search algorithm you have written and respond with:

```
{
  "found": true,
  "method": "oidc",
  "oidc_providers": ["google"] // if method = oidc
}
```

The method should probably be named something like `POST /credentials/known`.

What do you think?

File: identity/handler.go
Comment (aeneasr) [inline identity/handler.go:94]: This will panic because ok is checked afterwards. It might also be simpler to use `gjson.Get(...)` instead of marshalling / type asserting

File: identity/handler.go
Comment (aeneasr) [inline identity/handler.go:110]: This only takes the first provider but what if we have more than one?

File: identity/handler.go
Comment (aeneasr) [inline identity/handler.go:130]: `need-username` is not a method so I would just return `"method": "password"`

**Review by aeneasr — CHANGES_REQUESTED** [review]
Thank you for the contribution! Unfortunately we can not accept contributions without tests. These are important to ensure that your feature does what it should, and it ensures that it keeps working even if others change your code!

If you need help with the tests feel free to ping a maintainer!

File: identity/handler.go
Comment (jess-sheneberger) [reply identity/handler.go:75]: Sure, that seems fine, but how about this for the response?
```
{
    "found": true,
    "methods": [
        {
            "method": "password"
        },
        {
            "method": "oidc",
            "provider": "google"
        },
        {
            "method": "oidc",
            "provider": "microsoft"
        } // ...
    ]
}
```

because there could be multiple methods to login as well as multiple OIDC providers.

File: identity/handler.go
Comment (jess-sheneberger) [reply identity/handler.go:94]: Cool, thanks for the `gjson.Get()` suggestion, I will do it that way, definitely simpler

File: identity/handler.go
Comment (aeneasr) [reply identity/handler.go:75]: Big brain! 🧠 :) That makes total sense

File: identity/handler.go
Comment (aeneasr) [reply identity/handler.go:94]: I love that tool :)

**Comment by jess-sheneberger** [comment]
I believe I've incorporated all the review feedback now. I need a little help with writing some tests--It looks like this should be tested in handler_test.go, I see how to create identities but I don't see how to set credentials / associate with OIDC provider so I can exercise the new code. Can you point me in the right direction?

**Comment by jess-sheneberger** [comment]
Ping - can anyone help me get started writing the tests? I need to create a couple identities, one associated with OIDC and one with password credentials.

**Comment by zepatrik** [comment]
Sure, sorry for the late response from our side. I believe you can just use something similar to
https://github.com/ory/kratos/blob/103880c4fee41b1a6795eb031f921801891596f5/identity/handler_test.go#L180-L188
and
https://github.com/ory/kratos/blob/bce14c487450bd668859f362b98704644fa4c72a/identity/test/pool.go#L70-L74
and
https://github.com/ory/kratos/blob/7b8d59f48ed14a6d0672238645d8675d4bf7fd77/selfservice/strategy/oidc/strategy_settings_test.go#L114-L120

In general, these searches should give you more examples:
- https://github.com/ory/kratos/search?q=CredentialsTypeOIDC
- https://github.com/ory/kratos/search?l=Go&q=SetCredentials

**Comment by aeneasr** [comment]
While the PR is being worked on I will mark it as a draft. That declutters our review backlog :)

Once you're done with your changes and would like someone to review them, mark the PR as ready and request a review from one of the maintainers.

Thank you!

**Comment by jess-sheneberger** [comment]
Accidentally closed this, I will reopen a new PR and link this one when I'm finished with the tests.  Sorry about that, I'm still figuring out the Github workflow :-(

**Pull request closed without merging**
